Fix deferred start/broadcast of masternodes

domob1812 opened the pull request:
Deferred start of masternodes (with `startmasternode ALIAS true` and then `broadcaststartmasternode`) does not work properly if the `broadcaststartmasternode` RPC call happens on another node than the `startmasternode`.  While the broadcasting node does parse and verify the broadcast and even adds the masternode locally, it failed to insert the broadcast into `mapSeenMasternodeBroadcast`.  Hence it would send out an `inv` to its peers, but on request for the corresponding data, would not be able to find and relay it.

But going even further, there were three distinct places in the code for "starting masternodes":  `startmasternode`, `broadcaststartmasternode` and processing of received `mnb` messages.  All three of them should more or less do the same core logic, with only slight tweaks.  Thus, in this PR we unify most of the handling for all three into a new function `CMasternodeMan::ProcessBroadcast`.  This simplifies the code in general, and also makes sure to avoid subtle differences and bugs in the future.

Finally, the `mnoperation.py` regtest is extended to use deferred start for one of the two masternodes.  The modified test fails without the other code changes, and passes with them.

domob1812 committed:
Unify (broadcast)startmasternode code.

startmasternode and broadcaststartmasternode do something very similar,
namely processing and broadcasting the activation message for a masternode
(typically one owned by the person owning the node the RPC is run on as
well, but that's not specially required).

However, both routines used different code, and in particular
broadcaststartmasternode had a bug that didn't insert the broadcast
into mapSeenMasternodeBroadcast.  Thus the node would send out an
announcement of the broadcast message, but wouldn't be able to reply
to peers requesting the actual data (unless startmasternode with "defer"
and the subsequent broadcaststartmasternode would be done on the
same node).

This commit reduces code duplication by using a single code path to
process the masternode broadcast for both RPC methods (with the difference
that startmasternode first constructs the broadcast itself).  It also
extends the mnoperation.py regtest to use a deferred broadcast relay
for one of the two nodes to test it.

domob1812 committed:
Share code between mnb processing and register.

Registering a masternode (CActiveMasternode::Register) and processing
a "mnb" message from the network do very similar things (both simply
process a received mnb message, verify it and update our state
accordingly for it).

Thus we can share code, which also reduces the potential for bugs and
slight, unintended differences in behaviour.  (For instance, one such
difference was that the ping contained inside the broadcast message was
added to the list of known pings in one place and not the other.)

galpHub commented on divi/src/activemasternode.cpp:
```diff
@@ -171,27 +171,9 @@ bool CActiveMasternode::SendMasternodePing(std::string& errorMessage)
 
-    addrman.Add(CAddress(mnb.addr), CNetAddr("127.0.0.1"), 2 * 60 * 60);
-
-    auto mnp = mnb.lastPing;
-    mnodeman.mapSeenMasternodePing.insert(std::make_pair(mnp.GetHash(), mnp));
```
masternodeman's ProcessBroadcast does not in fact record the ping for the broadcast inside the mapSeenMasternodePing. Shouldn't it happen then after the function call (on success, of course)?

domob1812 replied:
Yes, this bit was one of the differences I noticed between the three places unified in this PR.  I thought about adding it to the new shared `ProcessBroadcast` function, but then went for not doing it (but with no particularly strong reason).

As far as my current understanding goes, adding the ping is not necessary, since we never actually relay its `inv` and thus also won't need to answer any queries for it from peers (instead it gets relayed and processed directly with the `mnb` message).  Furthermore, if it were necessary, then the old code would be broken as well, since only the node actually performing the start used to insert it into `mapSeenMasternodePing` (and not every node processing the broadcast message).

But I can also see the argument that it seems cleaner / safer / more straight-forward to add the ping in, since it is a ping we've seen after all.  In that case, we should add it to `ProcessBroadcast`.  Let me know if that's the option you prefer and I'm happy to do that.

galpHub replied:
I'm not advocating for moving the ping into the ProcessBroadcast function. More precisely, I'm merely indicating that during registration of a broadcast, the ping is recorded in order to avoid any confusion about the ping having been seen previously. Seems like it would lead to some communication confusion if I receive an unknown ping (that belongs to my masternode) but might somehow query my peers for it? Sounds like it would sort itself out but I don't like to make wagers like that haha

galpHub replied:
I think with that change this is ready to be merged

domob1812 replied:
I've done an update now that adds the ping to `mapSeenMasternodePing` in `ProcessBroadcast`.  This is still slightly different from the old behaviour, but I don't see why processing should differ for this detail between starting one's own masternode and processing a message from the network.  (But if you have a good reason, please let me know!)

galpHub reviewed (commented):
Very much like the direction that you are taking with these changes 👍

domob1812 committed:
Record only valid broadcasts/pings.

Masternode broadcasts and pings are referred to by their "hash", which
only includes some data fields.  This means that it is trivial to
construct a mnb or mnp message that has the same hash of some other one
but is invalid (e.g. by copying the original and invalidating the
signature).

Previously, the code would insert a mnb or mnp message into the map
of seen messages right when it received it.  This meant that someone
could forge those messages, and then peers receiving the invalid ones
would happily store them as "seen", ignore further messages with that
hash (e.g. the correct ones) and even send out the invalid ones when
asked for them by "inv".

With this change, we only mark messages as seen when they have actually
been validated and are correct.  This avoids all of these attack vectors.

Note that this means that someone can spam us with invalid messages, for
which we will keep running validation (and then disregard them).  This is
not a big issue, though, as each invalid message will increase the peer's
banscore, so that an attacker will eventually be disconnected before they
can DoS us meaningfully.

domob1812 commented:
I've also noticed another potential issue with the old code, which is fixed in ccae9da82436878a2d48482aaf78355b26af4de1.

domob1812 committed:
Remove dead code (UpdateMasternodeList).

CMasternodeMan::UpdateMasternodeList is dead code and not used
anywhere, so remove it.  (It duplicates some of the logic that is
now in ProcessBroadcast, which is the proper place for it.)

domob1812 committed:
Add masternode directly in ProcessBroadcast.

Instead of having a function CMasternodeBroadcast::CheckInputsAndAdd,
we now just have a read-only "CheckInputs" method.  The updates to
the state (in particular, adding the masternode to our list / removing
an existing other one) are done in CMasternodeMan::ProcessBroadcast,
where they belong together with some other updates we do anyway.

This also allows us to ensure that we add the broadcast message to
mapSeenMasternodeBroadcast before we actually relay the inv.

galpHub commented:
Hey @domob1812, I think this one is ready to merge. Could you fix the conflict with masternode.h? I've added some changes to allow for the timestamp for the ping to be the same as the timestamp for the broadcast to keep remote starts' timestamps consistent. Just want to make sure it works with the existing regtests before the merge

domob1812 committed:
Verify ping contained in mnb message.

This adds logic to verify the lastPing contained in mnb messages
when processing those.  Without it the ping itself could be malleated
by an attacker, and we would then happily save it in mapSeenMasternodePing
in the wrong version (and relay it from there).

To this end, we also refactor CMasternodePing::CheckAndUpdate to make
it independent of global state.  Instead of looking up and updating the
masternode from the global masternode manager, it is passed in so that
the method itself is not touching global state.  This allows us to
purely verify the ping when processing a broadcast (updating a temporary
CMasternode instance) and before changing any global state.

domob1812 commented:
I've updated this and rebased to latest `Development`.  I've now added logic to properly process the embedded ping in `mnb` messages, so that we won't accept invalid pings inside otherwise valid `mnb`'s (which would still open up attack surface).

To make this possible, I've also done two refactorings that separate "checking" from "updating" logic, so that we can really check everything first and only update the global state after:

1. Instead of `CheckInputsAndAdd`, we now have a read-only `CheckInputs` method and add the masternode explicitly after all validation is done.  This is much cleaner and easier to understand anyway.
2. Instead of accessing and updating global state from `CMasternodePing::CheckAndUpdate`, the method explicitly expects a `CMasternode` to be passed in, and then operates on that instance (not directly touching any global state).  This allows us to check the ping on a temporary `CMasternode` instance first, and only do the state updates at the very end if all is valid in `ProcessBroadcast`.

galpHub commented on divi/src/masternode.cpp:
```diff
@@ -276,9 +276,10 @@ bool CMasternode::UpdateFromNewBroadcast(CMasternodeBroadcast& mnb)
         lastTimeChecked = 0;
         int nDoS = 0;
-        if (mnb.lastPing == CMasternodePing() || (mnb.lastPing != CMasternodePing() && mnb.lastPing.CheckAndUpdate(nDoS, false))) {
+        if (mnb.lastPing != CMasternodePing() && mnb.lastPing.CheckAndUpdate(*this, nDoS, false)) {
             lastPing = mnb.lastPing;
```
`lastPing = mnb.lastPing`
Can be removed since it's already being changed in passing *this to the CheckAndUpdate method, no?

domob1812 replied:
Yes indeed, good catch.

galpHub commented on divi/src/masternode.cpp:
```diff
@@ -1048,79 +1049,83 @@ bool CMasternodePing::Sign(CKey& keyMasternode, CPubKey& pubKeyMasternode, bool
-            CMasternodeBroadcast mnb(*pmn);
-            uint256 hash = mnb.GetHash();
+            const uint256 hash = CMasternodeBroadcast(mn).GetHash();
             if (mnodeman.mapSeenMasternodeBroadcast.count(hash)) {
                 mnodeman.mapSeenMasternodeBroadcast[hash].lastPing = *this;
             }
 
```
This is modifying global state even if somewhere after this portion of the code the ping is deemed to have "failed to update", no?

domob1812 replied:
Oh yes, indeed.  I'll keep looking at this in more detail to decide how best to avoid this change as well (and do it properly later in the code).

domob1812 replied:
I've addressed this now, by doing the update both of `mapSeenMasternodePing` and the pings insider `mapSeenMasternodeBroadcast` with a new helper method, and using that method everywhere as needed after the validation.

The code in question itself has in turn been removed from `CheckAndUpdate`.

galpHub reviewed (commented):
Just a few minor items
